src/meta/src/hummock/manager/time_travel.rs

```diff
@@ -0,0 +1,482 @@
+            .filter(hummock_time_travel_delta::Column::VersionId.lte(actual_version_id))
+            .order_by_asc(hummock_time_travel_delta::Column::VersionId)
+            .all(&sql_store.conn)
+            .await?;
+        let mut actual_version = replay_archive(
+            replay_version.version.to_protobuf(),
+            deltas.into_iter().map(|d| d.version_delta.to_protobuf()),
+        );
+
+        let mut sst_ids = actual_version
+            .get_sst_ids()
+            .into_iter()
+            .collect::<VecDeque<_>>();
+        let sst_count = sst_ids.len();
+        let mut sst_id_to_info = HashMap::with_capacity(sst_count);
+        let sst_info_fetch_batch_size = std::env::var("RW_TIME_TRAVEL_SST_INFO_FETCH_BATCH_SIZE")
+            .unwrap_or_else(|_| "100".into())
+            .parse()
+            .unwrap();
+        while !sst_ids.is_empty() {
+            let sst_infos = hummock_sstable_info::Entity::find()
+                .filter(hummock_sstable_info::Column::SstId.is_in(
+                    sst_ids.drain(..std::cmp::min(sst_info_fetch_batch_size, sst_ids.len())),
+                ))
+                .all(&sql_store.conn)
+                .await?;
+            for sst_info in sst_infos {
+                let sst_info = sst_info.sstable_info.to_protobuf();
+                sst_id_to_info.insert(sst_info.sst_id, sst_info);
+            }
+        }
+        if sst_count != sst_id_to_info.len() {
+            return Err(Error::TimeTravel(anyhow!(format!(
+                "some SstableInfos not found for epoch {}, version {}",
+                query_epoch, actual_version_id,
+            ))));
+        }
+        refill_version(&mut actual_version, &sst_id_to_info);
+        Ok(actual_version)
+    }
+
+    pub(crate) async fn write_time_travel_metadata(
+        &self,
+        txn: &DatabaseTransaction,
+        version: Option<&HummockVersion>,
+        delta: HummockVersionDelta,
+        group_parents: &HashMap<CompactionGroupId, CompactionGroupId>,
+    ) -> Result<()> {
+        async fn write_sstable_infos(
+            sst_infos: impl Iterator<Item = &PbSstableInfo>,
+            txn: &DatabaseTransaction,
+        ) -> Result<usize> {
+            let mut count = 0;
+            for sst_info in sst_infos {
+                let m = hummock_sstable_info::ActiveModel {
+                    sst_id: Set(sst_info.sst_id.try_into().unwrap()),
+                    object_id: Set(sst_info.object_id.try_into().unwrap()),
+                    sstable_info: Set(sst_info.into()),
+                };
+                hummock_sstable_info::Entity::insert(m)
+                    .on_conflict(
+                        OnConflict::column(hummock_sstable_info::Column::SstId)
+                            .do_nothing()
+                            .to_owned(),
+                    )
+                    .do_nothing()
+                    .exec(txn)
```

Comment:
> an in-memory set for sst ids in hummock_sstable_info

+1 because it's in the critical path commit_epoch

> other time travel related meta store tables

It seems unnecessary to me because there won't be many conflict insert for them. Besides, for the read path the complete metadata is required rather than the index (*_id), which would be very large in memory.

Reply:
See `last_time_travel_snapshot_sst_ids`. It is much simpler than maintaining a full set of sst ids, at the cost of less accuracy, i.e. sometimes there's still attempt to write the same SST into meta store. But it's good enough and avoid introducing locking in `truncate_time_travel_metadata`.